### PR TITLE
Add support for IServiceProviderFactory

### PR DESCRIPTION
**Pull request by davidfowl**
- A new abstraction that allows the DI container author to describe how
  to translate an IServiceCollection to a specific container and then to an IServiceProvider.
- Hosting will use this abstraction to swap out the default container

https://github.com/aspnet/DependencyInjection/issues/442

/cc @pakrym @halter73

### CODE REVIEW

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/IServiceProviderFactory.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Provides and extension point for creating a container specific builder and an <see cref="IServiceProvider"/>.
```
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/IServiceProviderFactory.cs]: Provides an~~d~~ extension

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public abstract TContainerBuilder CreateBuilder(IServiceCollection services);
+
+        /// <summary>
+        /// Creates an <see cref="IServiceProvider"/> from the container builder.
+        /// </summary>
+        /// <param name="containerBuilder">The container builder</param>
+        /// <returns>An <see cref="IServiceProvider"/></returns>
+        public abstract IServiceProvider CreateServiceProvider(TContainerBuilder containerBuilder);
+
+        object IServiceProviderFactory.CreateBuilder(IServiceCollection services)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: `object IServiceProviderFactory.CreateBuilder(IServiceCollection services) => CreateBuilder(services)`

for :100: points.

**Comment by pranavkm** [comment]
Nice. :shipit: when everybody's happy.

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: Long line

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    /// Provides an abstract base for creating a container specific builder and an <see cref="IServiceProvider"/>.
+    /// </summary>
+    public abstract class ServiceProviderFactory<TContainerBuilder> : IServiceProviderFactory
+    {
+        /// <summary>
+        /// Creates a container builder from an <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The collection of services</param>
+        /// <returns>A container builder that can be used to create an <see cref="IServiceProvider"/>.</returns>
+        public abstract TContainerBuilder CreateBuilder(IServiceCollection services);
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: `TBuilder`? How are you supposed to get to this type from `IServiceProvider`? Right now you have to call this method and do `builder.GetType()` in order to get it, but that might not be the exact type you'd want. Say you have a `ServiceProviderFactory<IContainerBuilder>`, but the actual type is `ContainerBuilder`. How will you find the `ConfigureContainer` method? Check for the parameter type using `IsAssignableFrom`?

**Comment by khellang** [comment]
:shipit:

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs (hunk 3 above)
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: Yep, using the runtime type of the return value.

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Provides an abstract base for creating a container specific builder and an <see cref="IServiceProvider"/>.
+    /// </summary>
+    public abstract class ServiceProviderFactory<TContainerBuilder> : IServiceProviderFactory
```
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: Why is this required?

Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs]: It's not required. It's a convenient base class for people implementing it.

File: src/Microsoft.Extensions.DependencyInjection/DefaultServiceProviderFactory.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public class DefaultServiceProviderFactory : ServiceProviderFactory<IServiceCollection>
```
Comment (halter73) [inline src/Microsoft.Extensions.DependencyInjection/DefaultServiceProviderFactory.cs]: So this will be registered by hosting by default? I'm assuming we won't have an extension method like `WebHostBuilder.UseAutofac` for this.

Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection/DefaultServiceProviderFactory.cs]: > So this will be registered by hosting by default?

Yep

> I'm assuming we won't have an extension method like WebHostBuilder.UseAutofac for this.

Nope.

**Comment by pakrym** [comment]
:shipit:

**Comment by davidfowl** [comment]
Merged
